grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.grpc;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.netty.util.concurrent.EventExecutor;
+
+final class DeferredListener<I> extends ServerCall.Listener<I> {
+
+    private static final List<?> NOOP_TASKS = ImmutableList.of();
+
+    @Nullable
+    private final Executor blockingExecutor;
+    @Nullable
+    private final EventExecutor eventLoop;
+
+    private List<Consumer<Listener<I>>> pendingTasks = new ArrayList<>();
+    @Nullable
+    private Listener<I> delegate;
```

Comment:
It is double-checked by 
https://github.com/line/armeria/blob/576874c46a522ee9fb942e0a49230845042614fb/grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java#L158-L168
So it would be thread-safe eventually.

Reply:
I see, in that case could `delegate.onComplete();` be called twice?

Reply:
Very smart 😄  thanks for the explanation 👍